### ISSUE TITLE
Discussion: suitable pool of words

### ISSUE THREAD

**Issue by SirXOPHER** [SirXOPHER/Hangman#7]
The basic implementation of the game used a list of 100 random words for testing. The content has already been replaced with 100 nouns for a better player experience, but there seems to be much room for improvement. In order to find a suitable pool of words for the game, a discussion about the criteria might be helpful, leading to an appropriate size and quality of the words used for the game. At the moment, the size of the list is only 100 words. The only constraints are that every word is a noun and has more than 7 letters (to make it sufficiently interesting). The size, quality, constraints and source of the pool of words is up for debate.

Key questions:
- what is a good (minimal) size of the list?
- what are the characteristics of good/interesting words for the hangman game?
- should there be different levels of difficulty?
- what range of word length is the most fun and still fair (min/max)?
- could words be obtained from an online dictionary whilst applying appropriate constraints?

**Comment by albertopoljak** [comment]
Levels of difficulty sounds interesting - maybe add some uncommon words for harder difficulty. But this could be a problem since I don't think word list should be less than 1000 words and adding them by hand isn't practical. Maybe your idea of fetching words from online dictionary could solve this problem and it would be interesting to code it since I haven't done something like that before. Did you find any specific dictionaries/APIs?

And maybe we should have a .txt file of some 10 000 words that are auto generated (based on our criteria: nouns, min 7 chars etc.) for offline use for when the client can't connect to internet.

**Comment by SirXOPHER** [comment]
Hey Aberto! Great that you're still tuning in to this repository, and thanks for your input to kick off the discussion. You're right, manually curated lists are impractical, but fine-tuning of the selection criteria is critical to have a good quality of the pool of words and also when an implementation of levels of difficulty is considered. I fully agree with you about the minimal size of the pool.

My thoughts on the other key questions are: Thinking about the quality of nouns I think another constraint should be that no "proper names" (in the linguistic sense) are to be included, i.e. names of people, specific places and stuff. A minimal length of 7 or maybe 8 letters seems to work. A maximum could be useful to prevent awkwardly long words to appear in the game, perhaps something between 12 - 16 letters as an upper boundary?

I'm glad that you've picked up my idea to use an online dictionary. From what I saw by checking the big names in the dictionary industry, they are all keen on selling you some kind of data plan to use their API and they seem very vague about what usage looks like before registering an account. But maybe I wasn't looking hard enough. Anyway, I've found something in the non-profit educational realm. The Wordnik dictionary comes with a comprehensive API and they even offer some web tools where one can play around with the different functions, including a random words generator. The parameters seem to include almost all criteria for our fine-tuning, looks very promising:

[Wordnik API: Random Words](http://developer.wordnik.com/docs.html#!/words/getRandomWords_get_3)

The output is in JSON format, which I haven't processed in Java so far, but there are toolkits for that like [JSON.simple](https://code.google.com/archive/p/json-simple/) to handle the mapped output in Java. Signing up at Wordnik for a developer API key seems to be free of charge, but a bit of a headache is usually how to manage an API key inside a public repository, and moreover: there must be certain limits to what can be done with that free access. However, the output from the page where one can test the functions might be enough as it also generates a request URL, like [this one for example](http://api.wordnik.com:80/v4/words.json/randomWords?hasDictionaryDef=true&includePartOfSpeech=noun&excludePartOfSpeech=proper-noun&minCorpusCount=0&maxCorpusCount=-1&minDictionaryCount=1&maxDictionaryCount=-1&minLength=8&maxLength=16&limit=1000&api_key=a2a73e7b926c924fad7001ca3111acd55af2ffabf50eb4ae5), which basically generates a random words list using the demo API key. Again, looks very promising to me. But looking closer at the output, I can see some proper names, even though I set the parameter to exclude "proper-nouns". There are also terms consisting of more than one word, separated by either spaces or hyphens, so not very suitable for a game of Hangman, but that at least could be cleaned up while processing. I've also seen a ton of very exotic words, possibly increasing our difficulty level to an extremely frustrating point. But maybe I've not utilized the parameters appropriately.

Your idea about an offline fallback textfile (f.i. offline.txt) is great and should be implemented in case the online dictionary feature is feasible at all.

**Comment by albertopoljak** [comment]
I'm still waiting for the Wordnik API key. Meanwhile I found out that official links to java API don't work so I searched and found few repositories and already implemented one but can't test it properly because I still don't have the key :/

Also I have been thinking about offline word list and its generation and I'm thinking we should have more text files, one for each level of difficulty (if we even implement them). Since we are talking about thousands of words I have been searching for a way to auto generate them but failed to find suitable solution because all of the online services either generate 1 word or they generate words in an unreadable format (no save list  or no specific criteria of word generation like nouns) that would require extracting words from website then processing them to remove ordering/bullets and saving them to list. So until a find a service capable of generating dozen/hundreds of random words that fit my specific criteria (one word, noun etc.) I will not jump to coding extractor just jet. On the other hand I did find lots of already generated lists on the net, including sorted lists by nouns. Only problem is some words are too short (easily filtered by code) and some words are ... well too uncommon/hard.  Seems that auto generating a perfect list is quite hard. Maybe I write a small app that will help in creating list based on inputted words, then add options like removing words that don't fit criteria: too short / to long ,  already exists in list, change all chars to lowercase etc.... and then just copy pasting words from sorted list I found on net to that app. But that would still leave problem with words containing names or uncommon words. Maybe I compare created list to a list of names and remove those that match, same for uncommon words. 

These are just ideas I had while waiting for the key.

**Comment by SirXOPHER** [comment]
From what you wrote back, I take it that you do think the Wordnik API is worth a try. I was hoping you'd take a closer look, but didn't expect to get you all excited about it so promptly :) I did not yet apply for a key, as I was still pondering about the general feasability of our task, much in the way you did in your second paragraph...the "complexity" of the list indeed provides a massive headache!

For the very limited list, as it is used in our first working implemenation, I simply ran [this random noun generator](https://randomwordgenerator.com/noun.php). The result seems to contain no "proper names", but the word base looks very limited in scope, which is the opposite problem to what we face with the dictionary approach. Now I also realised that the links to the Java documentation for accessing the Wordnik API listed [on this page](http://developer.wordnik.com/#!/libraries) are broken, which is a shame! Did you mean that you went out and screened other repos where the API has been used, to learn from there how stuff works? That sounds exhausting. Since you invested some time already, it might be worth waiting for the key and give it a shot. I'm curious about your findings.

I'm absolutely with you regarding your other thoughts. It all boils down to the complexity issue of the generated word lists from the dictionary. It's a good idea to keep track of things we can fix and attend to by simple filtering in our code, but I'd also advise against starting to code this solution (extract, clean, reformat) before the more fundamental issue is resolved. The thought of using another dictionary search for cleansing purposes sounds interesting, but that might be computationally too intense for a client-side fetch of word lists. In this case we would have to go with a  pre-prepared offline list, or even several (if levels of difficulty are an option). Let's keep kicking such ideas around.

In summary: Our holy grail seems to be a large auto-generated random word list without "proper names". The subsequent filtering sounds not too complicated. If we can't get that, we could still get there using some sort of data cleansing, but in my view this might turn out as even harder to accomplish. But let's not drop it just because I think it's going to be painful ;)

**Comment by albertopoljak** [comment]
Huh turns out that Wordnik email service is "unreliable" and my API key has been in my account settings all along -.-  ,will get back at testing it in a few days and write my results here.

And yup I registered for the key simple because I wanted to test it and find out how exactly it worked plus I was curious about implementing other APIs. And finding that API was really minimum effort from my side , I simply googled "Wordnik Java API" and from few results choose the [simplest looking](https://github.com/jeremybrooks/knicker).

And yes , the thought of using another dictionary search for cleansing was meant to be one time only and specifically for cleaning offline list. The problem with cleansing would still remain with online services and I agree that that problem should be resolved by using their API criterias and not our offline cleaning methods.

Guess I'll play with Wordnik API before continuing my search for perfect noun list.

**Comment by SirXOPHER** [comment]
Haha, funny story. Always check your SPAM folder 😆 but seriously, I sometimes wonder about filter criteria in my email account, too.

About the repository you've found: Working with web service APIs is mostly new territory for me, so I didn't really get it first. Let me digest this slowly. Wordnik provides the data and in theory, they should also provide API definitions and documentation for a few common languages. They seemed to do that at some point, but the links to the Java API are broken (as we already found out earlier). That's where "Knicker" enters the stage. Some guy sat down and wrote an entire Java API for accessing the Wordnik database. The "Using knicker" section in the readme file describes a minimum working example of how to achieve that. Okay, the penny is dropping (slowly) 😉 

About the data cleaning process: I saw your earlier edit of the message in my emails that I get from GitHub. You were onto something with the database from Princeton University, but it didn't work out. Too bad, but this brings us back to sorting out the data retrieval from the Wordnik database. Speaking of which: I see a massive pull request which will import the Knicker API into our project and your associated comments. I'll comment over at the PR before we proceed.

Thanks for your commitment to this little sandbox!

**Comment by albertopoljak** [comment]
Only problem with license that I see is that there is no license ([link](https://github.com/jeremybrooks/knicker)) , but still we can't just dump all that code to ours. I think it would be fair just to make some note in our readme that we used his code, nothing special, just a mention.

And the directory problem, well I'm not exactly sure where our additional code should be based on [standard directory layout](https://maven.apache.org/guides/introduction/introduction-to-the-standard-directory-layout.html). For now I guess it should go in folder "it" meaning integration test. But that's a small problem and can be solved later. Also we are using many classes from Knicker that we do not need so we can remove it later after we test it out, this is also a minor issue.

And yea I still didn't find a perfect solution for our offline lists :/   , no biggie we'll eventually figure something out.

**Comment by SirXOPHER** [comment]
**License**: That's right, there is no `LICENSE.txt` or `LICENSE.md` in the original repo and still, there's talk about "_GNU General Public License_" within the Knicker code. This is a bit uncomfortable, but your proposal makes sense and in search for guidance I found this [sane-sounding advice](https://softwareengineering.stackexchange.com/a/304895). Let's make sure this external library always remains in a separate directory and I will update the readme file to our repo, as well as include the original readme file of Knicker and a GNU GPLv3 license file next to the code (which the Knicker creator expects us to generate...I assume).

**Directory structure**: I like that you bring up the [Standard Directory Layout](https://maven.apache.org/guides/introduction/introduction-to-the-standard-directory-layout.html) again, as I wanted to adopt some professional standards early on. I'm also a bit puzzled where the home of external libraries should be, but since the integration test folder is often used for plugins...why not? Maybe we'll find a better place, so let's shelve this issue for now, just like you suggested. I will open another issue as a reminder for a reorganisation of the project and to clean up what we don't use. However, we might have to keep the Knicker files as a whole, depending on whether our usage of Knicker is seen as a "redistribution" which surely requires to include all the files.

**Comment by SirXOPHER** [comment]
**Update**: I just created a license file for Knicker and put it in the same folder, next to the Knicker classes. At first I screwed up placing the file correctly and learned [moving a file to a new location](https://help.github.com/articles/moving-a-file-to-a-new-location/) on the GitHub website. I also updated the `README.md` within the `new-feature-dictionary` branch. The default branch is still `master` and whenever you enter the repo, it starts out at the default.

**Comment by SirXOPHER** [comment]
Unfortunately I don't have a lot of time for Hangman at the moment, but finally got your latest stuff working on my local machine. I recently changed IDE for my Java development, so had to figure out how to handle the Knicker code as none of the import statements worked at first. But now I can run the method calls you put together as tests. I can see some half-baked exception handling, but let's ignore it while we try to make the fundamental function work. The actual call of `randomWord()` looks pretty straight forward and I recognize all those parameters from [the "test drive" on the Wordnik developer website](http://developer.wordnik.com/docs.html#!/words/getRandomWords_get_3). In the 9args test (and less so in 12args) you do overwrite the result variable several times, but I assume this reflects your path in testing the functionality. I didn't fully understand all of the parameters and how they must be orchestrated to get closer to what we desire. What exactly is affected by the corpus count / frequency?

Example output from the 3 test methods:
`...0args()` : ferronickel
`...9args()` : sweep
`...12args()` : collarette, endotoxins, megaparsec

The random words in this example that are actual nouns would qualify all our formal requirements, but our concerns are pretty much the same we had after the test drive: way to exotic output of random words. I've heard of a 'parsec' before, but maybe only because I like science fiction a lot.

Alright, I will investigate parameter details and write some test methods of my own as soon as I can find another quiet hour. Just wanted to let you know about my method calls and first impressions. You were right, using Knicker is indeed way more straightforward than I initially had thought and it's a pleasure to see it running locally, but I wish there was a more detailed documentation.

**Comment by SirXOPHER** [comment]
**Update**: I just had a suspicion about the parameter set `minDictionaryCount` / `maxDictionaryCount` and tested some parameters using the Wordnik API website (convenient for a quick parameter test):
```
hasDictionaryDef=false
includePartOfSpeech=noun
excludePartOfSpeech=proper-noun
minCorpusCount=0
maxCorpusCount=-1
minDictionaryCount=30
maxDictionaryCount=-1
minLength=8
maxLength=12
```
By fine-tuning of `minDictionaryCount` the resulting random word lists include a lot more common words now and far less exotic ones! Parameters `minCorpusCount` and `maxCorpusCount` are still a mystery to me.

**Comment by albertopoljak** [comment]
I didn't check that out until now but the corpus frequency is:
"The corpus frequency is the measure of how many times a (case-sensitive) word has been found, and is based on the millions of documents consumed by Wordnik. You can use this measure to determine the "commonality" of a word." [link](https://stackoverflow.com/questions/11583339/what-is-minimum-corpus-frequency-for-terms-in-wordnik-api)
And second:
"minDictionaryCount and maxDictionaryCount simply corresponds to how many definitions the word has."

It's great that you found solution to complicated words and now we can even implement difficulty levels just by changing those 2 parameters. I'll have free time in a day or two so I'll test it some more.

**Comment by SirXOPHER** [comment]
Thanks for looking that one up. I must admit, I totally ignored those weird parameters at first, but a bit of research seems to pay out. However, I fail to get any beneficial effect from fiddling around with the corpus frequency parameter. Maybe I still don't get it or it's very subtle in consequence. It may require some more testing/research. The dictionary count parameter on the other hand, seems to provide good filter qualities. I tested a few different settings and my most apparent observation so far is: A higher `minDictionaryCount` setting gives us more common words, but also drastically shrinks the available words matching this criteria. This will require some more testing, so we can have a big enough pool with words one can actually guess.

I pushed up a new iteration of our `new-feature-dictionary` branch. I separated most Wordnik API related stuff from the Game class, just to tidy up a little. All your test methods can be found in the new class file and I added two more to start testing difficulty levels based on your method calls. To focus on one test at a time, every test method call is put into a comment, except the current one. I also added some line breaks and unnecessary print lines, just to make the test output more readable for my taste. Most of that decoration is just temporary 😉 

If we can balance out the parameters, we might have to write another filter method, because I think I saw some hyphen-saparated or space-separated words among my test results. Implementing a "coose a difficulty" prompt should be an easy ride. I'll be on a short trip abroad from Friday on, but will be back early next week.

**Comment by albertopoljak** [comment]
You should notice corpus frequency effects once you put some big numbers in minCorpusCount , I read that that value should be minimum 1000 if we want more common words. Maximum that I went for was 1M and then I only got a list of 16 words.

Also I tried putting multiple excludePartOfSpeech parameters (in fact all of them except nouns) but server returns error 500. So after a while of googling I found out that that parameter only works with ONE option and it's currently broken when including more of them. Basically to my current knowledge it is impossible to get a clean list of nouns.
Example: We want to include nouns but we will get the whole package including proper nouns, plural of nouns, posessive nouns etc. but we can only exclude ONE of those. 
But I did heard of some new update or some beta Wordnik that should resolve this problem so I'll investigate more.

**Comment by SirXOPHER** [comment]
I didn't try big numbers for `minCorpusCount` at first. Probably because I played with `minDictionaryCount` before and that was very sensitive, shrinking the dictionary to a few words at a value around 50. The corpus frequency seems to have the same effect towards the upper boundary. So it's really good that you found out about the threshold values for that parameter. My current guess is that a value around 10.000 might be viable for good measure.

Your other insight poses more of a problem for our progress. If there is indeed no way of excluding more categories from the result we are pretty much back to square one. I also found a Google user group where the author of Knicker [inquired about the very same issue](https://groups.google.com/forum/#!topic/wordnik-api/E2RXYsMyDVE), maybe that's what you were referring to?

I went back to the Wordnik Developer Website and tried to exclude multiple CSVs using their [basic web interface](http://developer.wordnik.com/docs.html#!/words/getRandomWords_get_3). The interface produces a "Request URL" (that's what I thought we might have to work with in the beginning) and I tried to simply insert more excluded values into the request being passed through variables in the URL: [This is the result](http://api.wordnik.com/v4/words.json/randomWords?hasDictionaryDef=false&includePartOfSpeech=noun&excludePartOfSpeech=proper-noun&excludePartOfSpeech=proper-noun-plural&excludePartOfSpeech=proper-noun-possesive&minCorpusCount=10000&maxCorpusCount=-1&minDictionaryCount=10&maxDictionaryCount=-1&minLength=8&maxLength=12&limit=1000&api_key=a2a73e7b926c924fad7001ca3111acd55af2ffabf50eb4ae5). It seems to work this way, but of course this would bypass Knicker alltogether and in order to use it, we would have to call the URL from within our program and process the JSON output. I'm not sure how messy this can get and surely, a fixed Knicker method call would be so much cleaner.

**Comment by albertopoljak** [comment]
I asked support and indeed there is a bug in java API that makes exclude parameter obsolete. I didn't know what to do so I even signed up for beta Wordnik. 
Your findings are truly ingenious, I somehow automatically assumed that even request URLs don't work but I tested your URLs and they seem to be working. I will take a look at JSON parsing and maybe even implement few test cases.

Nice work for thinking out of the box man.

Edit: While searching for JSON libraries it occurred to me...how come the request URL works and knicker API doesn't? So I did some testing and turns out that there was a bug in Knicker API (not in Wordnik API like developers have told me) where the request URL sent from Knicker methods was messed up in case there were more exclude parameters. Key problem is the function buildParamList in Util.java , 
seems that native function URLEncoder is having problems encoding input in the case of java.map having more string values for the same parameter. Anyway I have made new pull request where I completely scraped native Knicker code for method randomWords and instead of putting all that parameters in java.map I simply added them to String Builder and then passed that as String to get final request URL. Seems that Knicker guys failed to include case where you have multiple values for one parameter in java.map (although it's better than my solution) and therefore URLEncoder  gave all sort of weird characters in final request URL. I'm pretty sure I could have left their java.map idea but that would require more in depth analyzing. This fix works fine, although for other methods to work they would need to be altered in similiar fashion.

**Comment by SirXOPHER** [comment]
I'm glad my round of thinking directed you towards a possible solution. Once again, you propelled us forward with another piece of the puzzle (your latest PR). It's a bit confusing that this bug is commonly assumed to be a Wordnik API probem, when there seems to be a way around it from inside Knicker. I'm also glad that you kept on tracing this down, so we can continue using Knicker for the time being. This way we miss the opportunity to mess around with JSON, though. But I can live with that 😄 and there will be another time and use for that. Anyway, while researching the other day, I've found a [blog entry concerned with comparing the most popular JSON libraries for Java](http://blog.takipi.com/the-ultimate-json-library-json-simple-vs-gson-vs-jackson-vs-json/), if you are interested.

I took a closer look at your refactoring before merging the PR. I don't know exactly how Knicker does its thing, I've been treating it as a "black box", which is the idea when utilizing an API, isn't it? Until things don't work, of course 😆 Anyway, I just walked through your workaround and the `Util.java` class again to get a better understanding. Your explanation helped a lot to get a grasp of the situation. Not sure why this map structure is being used, but it might be the most versatile approach perhaps. I now understand what you did and I say we go with your workaround for now. If you feel the itch for a challenge, you can still come back and try to fix the general issue (multiple values per parameter via map structure). In fact, it would be quite an achievement to get something to work that left both Wordnik and Knicker developers puzzled for a couple of years.

Alright, I will go on and test our new capabilities this week, using your patched methods. If this works in all cases, we can soon leave `WordnikTest.java` behind (let's still keep it for testing) and add another class that produces all the random lists we need.

**Comment by albertopoljak** [comment]
I managed to fix the core problem some days ago. I made a [pull request in Knicker](https://github.com/jeremybrooks/knicker/pull/5). Therefore we can use the API without any modifications.
There is no need right now to use other libraries for JSON since we got Knicker up and running(unless you want to ofc), but I'm still not sure if that request URL that you found out works or not. It appears to be working but I'm still kinda suspicious ¯\_(ツ)_/¯

**Comment by SirXOPHER** [comment]
Now that's even better than your previous workaround (which you are rolling back with the latest PR). But the extra work paid off, as this is a more elegant solution to our problem. I hope the original Knicker repository is still being maintained and that your patch gets adopted.

No worries, I do not insist we play around with JSON 😀 when we now can use the API as intended. Much cleaner approach this way, brilliant! I'll pull it into the branch for use with the pending difficulty implementations. Looks like a big improvement.

**Comment by SirXOPHER** [comment]
_Hang in there, Hangman, we're coming to your rescue._

Well, I didn't have a lot of time lately, so I hope it is okay that I put this project into the slow cooker from time to time. Just picked it up again and refined the two candidate difficulty lists (easy + hard) a little more. Their new home is the class `WordPool.java` and I mixed parameters until I got 1000 random words back with each list. The printout for each list is deactivated (block comment) for now, as I went on and passed the lists into the game. The frustrating thing about it is, that I still saw some "proper nouns" coming up, like "Microsoft" or "Guelph" (a town in Canada) and the like. Your critical suspicion was well justified, as it appears. Also, there are occasionally some multi-part words, hyphenated or space separated. The good news is that I believe this can be fixed with a little bit of data cleaning. It appears that all "proper nouns" start with uppercase letters, whereas all other nouns start with lowercase letters. Filtering would need to purge all uppercase nouns, hyphenated and space separated words from the list. I'm not entirely sure those filter requirements are final, so I left this step for another session.

Over in the main game class I implemented a kind of switch that let's you coose the level of difficulty when you start up the game. I realised that logging in with the API key returns a Boolean (true, when sucessfully connected), so I used that to detect dictionary availability. The list used when offline is still the old textfile, which can also do with some improvement. I think there will be enough "waste" from testing and filtering the other lists. The template for the difficulty switch that I've used is your brilliant refinement of the user input prompt `playAgainPrompt()`, which is tasty and recursive(!). I tried something else with the switch `difficultyPrompt()` first, but again got runtime issues similar to those I had when we were starting out: user input simply didn't get recognised and so the program got jammed. I guess I need to revisit the Scanner class and its methods. Anyway, by following your code structure it works fine.

**Comment by albertopoljak** [comment]
Naah it's not a problem, that's a good thing about hobby , you can do it whenever you have time ;)
Btw I made a new pull request that has commits of its own :S
Also I wanted to ask ... if you were to download this program and paste it directly to your IDE , would it work out of the box? Without changing any variables like file paths etc.?

And yes, that word elimination would be easy to do. So then what's left to do? Offline word list, code clean-up and perhaps GUI?

**Comment by SirXOPHER** [comment]
Hey, thanks for your latest improvements, great stuff! More comments about that are right at the PR leading up to the most recent merge.

About testing hangman _taken right out of the shelve_ if I may say so. I'm using IntelliJ IDEA for development (only since very recently, so I'm not a pro) and it took some time to configure the project module correctly (i.e. so that it finds all dependencies, like the Knicker code). Since you asked, I picked up the old crutches and fired up BlueJ, downloaded our program from the branch we are working on as a ZIP and opened it as a non-BlueJ project, only to run into compile errors as it cannot find the location of the Knicker code (_"package net.jeremybrooks.knicker dies not exist"_). But BlueJ can be a bit tricky with implementing package structures, so I will give it another try to find out if this is the only issue. But the short answer is no, it doesn't work just like that. But source code behaves always a bit funny on different systems, right? Deployment is a different story, that's what JAR files are for, I think...at least that's the old fashioned way.

I also think we might slowly come to a conclusion with this particular issue. Our new feature is nearly complete, except some leftover work and polishing, yes. Dunno at what point one should build a "release" here on GitHub. Well, maybe not something as basic as this.

You mentioned a GUI. I have also thought about letting hangman escape from the console at some point. Great that you brought it up. I will research some options as a starting point and then open a new discussion issue. It would be great if you'd be up for it.

**Comment by albertopoljak** [comment]
Ayy, btw I asked that because I use Eclipse and I have some path issues which I have to change every time but I'm used to it and have a chunk of code ready to copy paste. I'll try to resolve that issue since it's on my side.

And GUI, well biggest issue with GUI implementation is implementing the editor in your IDE and even that is easy. After that it's a piece of cake. 
I already have experience with JAVA SWING (Eclipse editor is called [window builder pro](http://download.eclipse.org/windowbuilder/WB/integration/4.4/) , find something for IntelliJ ), but it looks somehow roboust and old fashioned. But if that's the problem I found a great repo the other day [link](https://github.com/mgarin/weblaf), it looks very fancy. Although I'm not sure if it's necessary to be that fancy since we are going to have very little GUI elements.

And about that merge comment:
 I guess I switched to ArrayList because I was testing something ¯\_(ツ)_/¯ and I forgot to change it back because it was on the top so I didn't see,  we should go back to List.
And I removed the throw declaration because nothing was catching the exception and besides exception is handled inside the method.
